File: lib/pipeline.ts
/**
 * Declarative Pipeline Engine
 *
 * Pure functional composition primitives for immutable data transformations.
 * All operations are referentially transparent — no side effects, no mutation.
 *
 * Complexity: pipe/compose are O(n) where n = number of stages.
 * Memory: each stage produces a new immutable value; no in-place mutation.
 */

/** A pure transform: takes A, returns B. No side effects. */
export type Transform<A, B> = (input: A) => B;

/** A predicate: takes A, returns boolean. */
export type Predicate<A> = (input: A) => boolean;

/** Result of a validated pipeline stage — either success or failure with reason. */
export type PipelineResult<T> =
  | { readonly ok: true; readonly value: T }
  | { readonly ok: false; readonly error: string };

/** Compose two transforms left-to-right: f then g. */
export function compose<A, B, C>(
  f: Transform<A, B>,
  g: Transform<B, C>
): Transform<A, C> {
  return (input: A) => g(f(input));
}

/** Pipe a value through an ordered array of transforms. O(n) sequential. */
export function pipe<T>(initial: T, ...stages: ReadonlyArray<Transform<T, T>>): T {
  return stages.reduce<T>((acc, stage) => stage(acc), initial);
}

/** Lift a pure function into a PipelineResult chain. */
export function lift<A, B>(
  transform: Transform<A, B>
): Transform<PipelineResult<A>, PipelineResult<B>> {
  return (input) => {
    if (!input.ok) return input;
    try {
      return { ok: true, value: transform(input.value) };
    } catch (err) {
      return { ok: false, error: err instanceof Error ? err.message : String(err) };
    }
  };
}

/** Guard: validate a value against a predicate, short-circuit on failure. */
export function guard<T>(
  predicate: Predicate<T>,
  errorMessage: string
): Transform<PipelineResult<T>, PipelineResult<T>> {
  return (input) => {
    if (!input.ok) return input;
    return predicate(input.value)
      ? input
      : { ok: false, error: errorMessage };
  };
}

/** Create a successful PipelineResult. */
export function succeed<T>(value: T): PipelineResult<T> {
  return { ok: true, value };
}

/** Create a failed PipelineResult. */
export function fail<T = never>(error: string): PipelineResult<T> {
  return { ok: false, error };
}

/** Chain multiple result-aware transforms. */
export function chain<T>(
  initial: PipelineResult<T>,
  ...stages: ReadonlyArray<Transform<PipelineResult<T>, PipelineResult<T>>>
): PipelineResult<T> {
  return stages.reduce<PipelineResult<T>>((acc, stage) => stage(acc), initial);
}

/** Map over a readonly array with a pure transform. Returns new array. */
export function mapImmutable<A, B>(
  items: ReadonlyArray<A>,
  transform: Transform<A, B>
): ReadonlyArray<B> {
  return items.map(transform);
}

/** Filter a readonly array with a predicate. Returns new array. */
export function filterImmutable<A>(
  items: ReadonlyArray<A>,
  predicate: Predicate<A>
): ReadonlyArray<A> {
  return items.filter(predicate);
}

/** Partition a readonly array into [matches, nonMatches]. */
export function partition<A>(
  items: ReadonlyArray<A>,
  predicate: Predicate<A>
): readonly [ReadonlyArray<A>, ReadonlyArray<A>] {
  const matches: A[] = [];
  const nonMatches: A[] = [];
  for (const item of items) {
    (predicate(item) ? matches : nonMatches).push(item);
  }
  return [matches, nonMatches] as const;
}

/** Aggregate/reduce a readonly array to a single value. */
export function fold<A, B>(
  items: ReadonlyArray<A>,
  initial: B,
  reducer: (acc: B, item: A) => B
): B {
  return items.reduce(reducer, initial);
}

File: runtime/pipeline.py
"""
Declarative Pipeline Engine (Python Runtime)

Pure functional composition primitives for immutable data transformations.
All operations are referentially transparent with no side effects.

Complexity: pipe/compose are O(n) where n = number of stages.
Memory: each stage produces a new immutable value; no in-place mutation.
"""

from __future__ import annotations

from dataclasses import dataclass
from functools import reduce
from typing import Any, Callable, Generic, List, Optional, Tuple, TypeVar

A = TypeVar("A")
B = TypeVar("B")
C = TypeVar("C")
T = TypeVar("T")


@dataclass(frozen=True)
class Result(Generic[T]):
    """Discriminated union: either success with value or failure with error."""

    ok: bool
    value: Optional[T]
    error: Optional[str]

    @staticmethod
    def succeed(value: T) -> "Result[T]":
        return Result(ok=True, value=value, error=None)

    @staticmethod
    def fail(error: str) -> "Result[Any]":
        return Result(ok=False, value=None, error=error)


def compose(f: Callable[[A], B], g: Callable[[B], C]) -> Callable[[A], C]:
    """Compose two transforms left-to-right: f then g."""

    def composed(x: A) -> C:
        return g(f(x))

    return composed


def pipe(initial: T, *stages: Callable[[T], T]) -> T:
    """Pipe a value through an ordered sequence of transforms. O(n) sequential."""
    return reduce(lambda acc, stage: stage(acc), stages, initial)


def lift(transform: Callable[[A], B]) -> Callable[[Result[A]], Result[B]]:
    """Lift a pure function into a Result chain."""

    def lifted(result: Result[A]) -> Result[B]:
        if not result.ok:
            return Result.fail(result.error or "unknown error")
        try:
            return Result.succeed(transform(result.value))  # type: ignore[arg-type]
        except Exception as exc:
            return Result.fail(str(exc))

    return lifted


def guard(predicate: Callable[[T], bool], error_message: str) -> Callable[[Result[T]], Result[T]]:
    """Guard: validate a value against a predicate, short-circuit on failure."""

    def guarded(result: Result[T]) -> Result[T]:
        if not result.ok:
            return result
        if not predicate(result.value):  # type: ignore[arg-type]
            return Result.fail(error_message)
        return result

    return guarded


def chain(initial: Result[T], *stages: Callable[[Result[T]], Result[T]]) -> Result[T]:
    """Chain multiple result-aware transforms."""
    return reduce(lambda acc, stage: stage(acc), stages, initial)


def map_immutable(items: List[A], transform: Callable[[A], B]) -> List[B]:
    """Map over a list with a pure transform. Returns new list."""
    return [transform(item) for item in items]


def filter_immutable(items: List[A], predicate: Callable[[A], bool]) -> List[A]:
    """Filter a list with a predicate. Returns new list."""
    return [item for item in items if predicate(item)]


def partition(items: List[A], predicate: Callable[[A], bool]) -> Tuple[List[A], List[A]]:
    """Partition a list into (matches, non_matches)."""
    matches: List[A] = []
    non_matches: List[A] = []
    for item in items:
        (matches if predicate(item) else non_matches).append(item)
    return (matches, non_matches)


def fold(items: List[A], initial: B, reducer: Callable[[B, A], B]) -> B:
    """Aggregate/reduce a list to a single value."""
    return reduce(reducer, items, initial)
